gapis/resolve/dependencygraph2/builder.go

```diff
@@ -0,0 +1,303 @@
+	currentDependencyExists []bool
+
+	// currentDependencies is a slice containing all of the nodes depended on by currentNodeID
+	currentDependencies []NodeID
+
+	// currentSubCmdIdx stores the SubCmdIdx of the latest effect
+	// currentSubCmdIdx api.SubCmdIdx
+
+	// currentCreateCount is a count of how many CreateResource effects are associated with currentNodeID
+	currentCreateCount int
+
+	// fragmentWrites stores all non-overwritten writes to the state, excluding memory writes.
+	// fragmentWrites[ref][frag] gives the latest node to write to the fragment
+	// frag in the object corresponding to ref
+	fragmentWrites map[api.RefID]map[api.Fragment]NodeID
+
+	// memoryWrites stores all non-overwritten writes to memory ranges
+	memoryWrites map[memory.PoolID]*memoryWriteList
+
+	// openForwardDependencies tracks pending forward dependencies, where the
+	// first node has been processed, but the second node has not yet been
+	// processed. The keys are the unique identifier for the forward
+	// dependency.
+	openForwardDependencies map[interface{}]NodeID
+
+	Stats struct {
+		NumCmdNodes     uint64
+		NumObsNodes     uint64
+		NumReads        uint64
+		NumWrites       uint64
+		NumMemoryReads  uint64
+		NumMemoryWrites uint64
+	}
+}
+
+// Get the graph which is being built.
+func (b *dependencyGraphBuilder) Graph() DependencyGraph {
+	return b.graph
+}
+
+// Build a new DependencyGraphBuilder.
+func NewDependencyGraphBuilder(ctx context.Context, config DependencyGraphConfig,
+	c *capture.Capture, initialCmds []api.Cmd, initialCmdRanges interval.U64RangeList) DependencyGraphBuilder {
+	graph := newDependencyGraph(ctx, config, c, initialCmds, initialCmdRanges)
+	builder := &dependencyGraphBuilder{
+		graph: graph,
+		currentDependencyExists: make([]bool, graph.NumNodes()),
+		currentCmdID:            api.CmdNoID,
+		currentNodeID:           NodeNoID,
+		fragmentWrites:          make(map[api.RefID]map[api.Fragment]NodeID),
+		memoryWrites:            make(map[memory.PoolID]*memoryWriteList),
+		openForwardDependencies: make(map[interface{}]NodeID),
+	}
+	return builder
+}
+
+// BeginCmd is called by the API emulator at the beginning of each API call
+func (b *dependencyGraphBuilder) OnBeginCmd(ctx context.Context, cmdID api.CmdID, cmd api.Cmd) {
+	debug(ctx, "OnBeginCmd [%d] %v", cmdID, cmd)
+	b.currentCmdID = cmdID
+	b.currentCmd = cmd
+	b.currentNodeID = b.graph.GetNodeID(CmdNode{api.SubCmdIdx{uint64(cmdID)}})
+}
+
+// EndCmd is called by the API emulator at the end of each API call
+func (b *dependencyGraphBuilder) OnEndCmd(ctx context.Context, cmdID api.CmdID, cmd api.Cmd) {
+	debug(ctx, "OnEndCmd [%d] %v", cmdID, cmd)
+	if len(b.currentDependencies) > 0 {
+		sortedDeps := make([]NodeID, len(b.currentDependencies), len(b.currentDependencies)+b.currentCreateCount)
+		copy(sortedDeps, b.currentDependencies)
+		b.currentDependencies = b.currentDependencies[:0]
+		sort.Slice(sortedDeps, func(i, j int) bool { return sortedDeps[i] < sortedDeps[j] })
```

Comment:
Do we really need to sort here?

Reply:
You're correct that the sort is not currently needed. The sort will be needed if we want to query whether one node is a dependency of another node. I will take the sort out for now, and put it back if/when it is actually needed.